CMP1903M Assessment 1 Base Code/Analyse.cs

```diff
@@ -1,54 +1,134 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.IO;
 using System.Linq;
 using System.Text;
 using System.Text.RegularExpressions;
 using System.Threading.Tasks;
 
 namespace CMP1903M_Assessment_1_Base_Code
 {
-    public static class Analyse
+    /// <summary>
+    /// Analyses text
+    /// </summary>
+    public class Analyse
     {
-        //Handles the analysis of text
+        private string _analysisText;
+        /// <summary>
+        /// Text to analyse, runs AnalyseText method when set
+        /// </summary>
+        public string AnalisisText
+        {
+            get { return _analysisText; }
+            set
+            {
+                _analysisText = value;
+                AnalyseText(_analysisText);
+            }
+        }
+        
+        /// <summary>
+        /// Keys as text elements, values as count
+        /// </summary>
+        public Dictionary<string, int> Elements { get; private set; }
+        /// <summary>
+        /// Keys as characters, values as count
+        /// </summary>
+        public Dictionary<string, int> Characters { get; private set; }
+
+        public Analyse()
+        {
+            
+        }
+        /// <summary>
+        /// Runs analysis methods
+        /// </summary>
+        /// <param name="input">Text to analyse</param>
+        protected void AnalyseText(string input)
+        {
+            Elements = CountElements(input);
+            Characters = CountChars(input);
+        }
 
-        //Method: analyseText
-        //Arguments: string
-        //Returns: list of integers
-        //Calculates and returns an analysis of the text
-        public static Dictionary<string, int> AnalyseText(string input)
+        /// <summary>
+        /// Analyses text using instances of AnalysisTool
+        /// </summary>
+        /// <param name="text">Text to analyse</param>
+        /// <returns>dict: Keys as text elements and values corresponding to count</returns>
+        protected Dictionary<string, int> CountElements(string text)
         {
             //List of integers to hold the first five measurements:
             //1. Number of sentences
             //2. Number of vowels
             //3. Number of consonants
             //4. Number of upper case letters
             //5. Number of lower case letters
-            //Initialise all the values in the list to '0'
+            
+            // Create dictionary with anonymous functions that count elements using Regex
             var tools = new Dictionary<string, Func<string, int>>()
             {
-                {"sentences", s => Regex.Matches(s, @"(\s|^)+[^.!?]*([.!?]|\S)", RegexOptions.Multiline).Count()},
+                {"sentences", s => Regex.Matches(s, @"(\s|^)+[^.!?\*]*([.!?\*]|\S)", RegexOptions.Multiline).Count()},

```

Comment:
This could be made simpler. Try not to reinvent the wheel!
```suggestion
                {"sentences", s => Regex.Matches(s, @"[.!?]\s", RegexOptions.Multiline).Count + 1,
```
This works as long as you make sure that the is no whitespace, line break or otherwise, at the end of the input string.

Reply:
True, however, this won't catch all edge cases. For example, if the user enters a sentence without ending punctuation, I would argue that input such as "This is a sentence" should still be counted as a sentence. Additionally, if there is no whitespace after the input, as in "This is a sentence." the group will not be caught also.

Reply:
That is by design.
The regex match statement is designed to catch all sentences besides the last one.
The last sentence which is instead counted by the +1 (since you can't have 0 sentences to begin with).
The reason you need to make sure there is no whitespace at the end of the input is so that the regex statement doesn't count the last sentence.

Reply:
Ah ok, that makes more sense now. With a guard clause stopping empty input and trimming end whitespace to make sure the last sentence is not counted, that would make the regex simpler. I'll implement this, thanks 👍